### PR TITLE
8294850: Make rs length/pending card predictors dependent on gc phase

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that separates `rs_length` and `pending_cards` predictors by young phase?

Some notes:
* this does not intend to improve the predictors themselves - i.e. the use of `xxx_diff` and `xxx` predictors and adding them together is kind of weird, imho trying to add a kind of prediction compensation/improvement to act on differences like a (PID-)controller. I also strongly suspect that the `xxx_diff` predictors ought to be allowed to report negative values. This is no attempt to fix this
* the parameter `for_young_gc` in the analytics API is misnamed. It's not about young gcs, but young_only_phase - I will fix this in a follow-up

Depends on PR#10649, please review that one first.

Thanks, 
  Thomas

Testing: gha, local testing, tier1-5 with other changes in this patch series, perf testing with other changes in this patch series

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294850](https://bugs.openjdk.org/browse/JDK-8294850): Make rs length/pending card predictors dependent on gc phase

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [35e6be38](https://git.openjdk.org/jdk/pull/10650/files/35e6be383f30659e792258c9a459039c210ffdfe)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [35e6be38](https://git.openjdk.org/jdk/pull/10650/files/35e6be383f30659e792258c9a459039c210ffdfe)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [5b6a3f93](https://git.openjdk.org/jdk/pull/10650/files/5b6a3f938b69614e7d329fad3e45cba20621c530)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10650/head:pull/10650` \
`$ git checkout pull/10650`

Update a local copy of the PR: \
`$ git checkout pull/10650` \
`$ git pull https://git.openjdk.org/jdk pull/10650/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10650`

View PR using the GUI difftool: \
`$ git pr show -t 10650`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10650.diff">https://git.openjdk.org/jdk/pull/10650.diff</a>

</details>

### CODE REVIEW

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by walulyai — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1Analytics.cpp [hunk 1]
```diff
@@ -301,12 +316,20 @@ double G1Analytics::predict_cleanup_time_ms() const {
...
+    return predict_size(_young_rs_length_seq) + predict_size(_young_rs_length_diff_seq);
+  } else {
+    return predict_size(_mixed_rs_length_seq) + predict_size(_mixed_rs_length_diff_seq);
+  }
 }
 
-size_t G1Analytics::predict_pending_cards() const {
-  return predict_size(_pending_cards_seq);
+size_t G1Analytics::predict_pending_cards(bool for_young_gc) const {
+  if (for_young_gc || !enough_samples_available(_young_pending_cards_seq)) {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1Analytics.cpp:328]: Why `_young_pending_cards_seq`? I'd expect `_mixed_rs_length_seq` just to match its neighbor.

**Review by albertnetymk — COMMENTED** [review]
`diff/pending_cards/rs_length` are more or less duplicated for young/mixed. I wonder if it's cleaner to encapsulate them in one data structure.

File: src/hotspot/share/gc/g1/g1Analytics.cpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1Analytics.cpp:328]: Typo from breaking up the big patch I had. Good catch! The correct sequence is `_mixed_pending_cards_seq` though.

**Comment by tschatzl** [comment]
> `diff/pending_cards/rs_length` are more or less duplicated for young/mixed. I wonder if it's cleaner to encapsulate them in one data structure.

I also considered this, but `rs_length_diff` will go away completely with other changes after the refinement ones, not sure if `pending_cards` separation will stay with Kim's changes. Then only `rs_length` is left.

There are a few others though already. What about doing that separately as cleanup?

**Review by albertnetymk — APPROVED** [review]

**Comment by tschatzl** [comment]
Thanks @albertnetymk @walulyai @kimbarrett for your reviews
/integrate

**Pull request closed without merging**